File: services/federacion/src/aqyra_federacion/bcf.py
"""emitir_bcf(informe, carpeta) → contenedor BCF 3.0 (un topic por incidencia).

BCF 3.0 de buildingSMART POR REFERENCIA (D2: no se re-esquematiza; la versión se
ancla en versions.lock [contracts.C4] bcf="3.0"). La emisión es un paso SEPARADO
de validar() (D11): la salida por defecto del service sigue declarando
bcf.emitido=false (D8) y el expected de C4-FED-01 queda intocado por construcción.

Determinismo — garantía del contrato (D13):
- GUID de topic     = uuid5(NAMESPACE_AQYRA, f"{caso}/{incidencia.id}")
- GUID de viewpoint = uuid5(NAMESPACE_AQYRA, f"{caso}/{incidencia.id}/viewpoint")
  con NAMESPACE_AQYRA = uuid5(NAMESPACE_DNS, "aqyra.bcf"): estable entre
  ejecuciones y ANCLABLE en la golden (C4-FED-02).
- autor/fecha son metadatos de generación INYECTABLES (misma filosofía que la
  `procedencia` del manifiesto); la golden los fija en su expected
  (bcf_generacion) y el runner los inyecta en el recompute.

Contenedor — el ÁRBOL descomprimido es lo que se ancla (D12, md5 por fichero):
    <carpeta>/bcf.version                  VersionId 3.0
    <carpeta>/<topic-guid>/markup.bcf      Topic: Guid, TopicType=Issue,
                                           TopicStatus=Open, Title, Priority ←
                                           severidad (literal), Labels =
                                           requisito + modelo:<id>,
                                           CreationDate/Author, Description
    <carpeta>/<topic-guid>/viewpoint.bcfv  Components → GUIDs afectados. Solo si
                                           la incidencia trae GUIDs (las reglas de
                                           módulo a nivel de proyecto emiten topic
                                           sin viewpoint). CÁMARA perspectiva
                                           determinista SOLO si emitir_bcf recibe
                                           `derivado=` (D29 — la cámara llegó con
                                           el IFC federado derivado, v0.x/D26;
                                           sin derivado, byte a byte como en v0:
                                           C4-FED-02/04 intactos por construcción).
El `.bcfzip` de intercambio es un DERIVADO sin anclar (un zip no es determinista
byte a byte); lo genera el CLI con --bcfzip.
"""
from __future__ import annotations

import copy
import uuid
from datetime import date
from pathlib import Path
from xml.sax.saxutils import escape, quoteattr

NAMESPACE_AQYRA = uuid.uuid5(uuid.NAMESPACE_DNS, "aqyra.bcf")

_DECL = '<?xml version="1.0" encoding="utf-8"?>\n'
_VERSION_XML = _DECL + '<Version VersionId="3.0"/>\n'


def guid_topic(caso: str, incidencia_id: str) -> str:
    return str(uuid.uuid5(NAMESPACE_AQYRA, f"{caso}/{incidencia_id}"))


def guid_viewpoint(caso: str, incidencia_id: str) -> str:
    return str(uuid.uuid5(NAMESPACE_AQYRA, f"{caso}/{incidencia_id}/viewpoint"))


def _markup(inc: dict, topic_guid: str, vp_guid: str | None,
            autor: str, fecha: str) -> str:
    """markup.bcf (BCF 3.0): orden de elementos del XSD de buildingSMART
    (Title, Priority?, Labels?, CreationDate, CreationAuthor, Description?,
    Viewpoints?). Priority ← severidad LITERAL; el requisito y el modelo van
    como Labels (referencia al requisito sin re-esquematizar nada)."""
    L = [_DECL, "<Markup>\n",
         f'  <Topic Guid={quoteattr(topic_guid)} TopicType="Issue" TopicStatus="Open">\n',
         f"    <Title>{escape(inc['titulo'])}</Title>\n",
         f"    <Priority>{escape(inc['severidad'])}</Priority>\n",
         "    <Labels>\n",
         f"      <Label>{escape(inc['requisito'])}</Label>\n",
         f"      <Label>modelo:{escape(inc['modelo'])}</Label>\n",
         "    </Labels>\n",
         f"    <CreationDate>{escape(fecha)}</CreationDate>\n",
         f"    <CreationAuthor>{escape(autor)}</CreationAuthor>\n",
         f"    <Description>Requisito {escape(inc['requisito'])} no conforme "
         f"en el modelo {escape(inc['modelo'])}.</Description>\n"]
    if vp_guid:
        L += ["    <Viewpoints>\n",
              f"      <ViewPoint Guid={quoteattr(vp_guid)}>\n",
              "        <Viewpoint>viewpoint.bcfv</Viewpoint>\n",
              "      </ViewPoint>\n",
              "    </Viewpoints>\n"]
    L += ["  </Topic>\n", "</Markup>\n"]
    return "".join(L)


def _xyz(tag: str, v: tuple, indent: str) -> str:
    """Tríada X/Y/Z del visinfo.xsd, floats a 6 decimales FIJOS (D29: el md5 del
    árbol BCF no puede depender del repr() de un float)."""
    i2 = indent + "  "
    return (f"{indent}<{tag}>\n"
            f"{i2}<X>{v[0]:.6f}</X>\n{i2}<Y>{v[1]:.6f}</Y>\n{i2}<Z>{v[2]:.6f}</Z>\n"
            f"{indent}</{tag}>\n")


def _viewpoint(vp_guid: str, guids: list[str], camara: dict | None = None) -> str:
    """viewpoint.bcfv (BCF 3.0): Components/Selection con los GUIDs afectados.

    `camara`: PerspectiveCamera determinista (D29, calculada por
    derivar.camara_para_guids sobre el IFC federado DERIVADO) — solo cuando la
    emisión recibe el derivado; sin ella el fichero es byte a byte el de v0 (D6)."""
    L = [_DECL, f"<VisualizationInfo Guid={quoteattr(vp_guid)}>\n",
         "  <Components>\n", "    <Selection>\n"]
    L += [f"      <Component IfcGuid={quoteattr(g)}/>\n" for g in guids]
    L += ["    </Selection>\n",
          '    <Visibility DefaultVisibility="true"/>\n',
          "  </Components>\n"]
    if camara:
        L += ["  <PerspectiveCamera>\n",
              _xyz("CameraViewPoint", camara["posicion"], "    "),
              _xyz("CameraDirection", camara["direccion"], "    "),
              _xyz("CameraUpVector", camara["arriba"], "    "),
              f"    <FieldOfView>{camara['fov_deg']:.6f}</FieldOfView>\n",
              f"    <AspectRatio>{camara['aspecto']:.6f}</AspectRatio>\n",
              "  </PerspectiveCamera>\n"]
    L += ["</VisualizationInfo>\n"]
    return "".join(L)


def emitir_bcf(informe: dict, carpeta: Path, caso: str | None = None,
               autor: str | None = None, fecha: str | None = None,
               derivado: Path | None = None) -> dict:
    """Emite el contenedor BCF 3.0 del informe QA: un topic por incidencia.

    `carpeta` ES el contenedor (bcf.version en su raíz). Devuelve un informe
    NUEVO (no muta el de entrada) con bcf = {…, emitido: true, carpeta:
    <nombre>} y `bcf_topic_guid` en cada incidencia — el informe solo refleja
    emitido=true cuando la emisión ocurre de verdad.

    `caso`: semilla de los GUIDs deterministas (por defecto, informe.proyecto).
    `autor`/`fecha`: metadatos de generación (por defecto, el service y hoy).
    `derivado`: ruta del IFC federado DERIVADO (D26); si se pasa, cada viewpoint
    gana la CÁMARA determinista de sus GUIDs (D29). Sin él, sin cámara (v0).
    """
    carpeta = Path(carpeta)
    if caso is None:
        caso = informe["proyecto"]
    if autor is None:
        from . import __version__
        autor = f"services/federacion {__version__}"
    if fecha is None:
        fecha = f"{date.today().isoformat()}T00:00:00Z"

    ifc_derivado = None
    if derivado is not None:
        from .lectura import abrir_ifc
        ifc_derivado = abrir_ifc(Path(derivado))

    carpeta.mkdir(parents=True, exist_ok=True)
    (carpeta / "bcf.version").write_text(_VERSION_XML, encoding="utf-8", newline="\n")

    out = copy.deepcopy(informe)
    for inc in out.get("incidencias", []):
        tg = guid_topic(caso, inc["id"])
        vg = guid_viewpoint(caso, inc["id"]) if inc.get("guids") else None
        d = carpeta / tg
        d.mkdir(exist_ok=True)
        (d / "markup.bcf").write_text(_markup(inc, tg, vg, autor, fecha),
                                      encoding="utf-8", newline="\n")
        if vg:
            camara = None
            if ifc_derivado is not None:
                from .derivar import camara_para_guids
                camara = camara_para_guids(ifc_derivado, inc["guids"])
            (d / "viewpoint.bcfv").write_text(_viewpoint(vg, inc["guids"], camara),
                                              encoding="utf-8", newline="\n")
        inc["bcf_topic_guid"] = tg

    out["bcf"] = {"estandar": "BCF",
                  "version": str(informe.get("bcf", {}).get("version", "3.0")),
                  "emitido": True,
                  "carpeta": carpeta.name}
    return out


def empaquetar_bcfzip(carpeta: Path, destino: Path) -> Path:
    """DERIVADO de intercambio (D12): zip del contenedor. NO se ancla en la
    golden (un zip no es determinista byte a byte: timestamps/orden)."""
    import zipfile
    carpeta, destino = Path(carpeta), Path(destino)
    with zipfile.ZipFile(destino, "w", zipfile.ZIP_DEFLATED) as z:
        for p in sorted(carpeta.rglob("*")):
            if p.is_file():
                z.write(p, p.relative_to(carpeta).as_posix())
    return destino

File: apps/visor/src/bcf.ts
/** Lectura BCF 3.0 (buildingSMART) — markup.bcf + viewpoint.bcfv.
 *
 * El visor CONSUME el árbol BCF que emite services/federacion (D12: el árbol
 * descomprimido es el artefacto; D29: PerspectiveCamera determinista en el
 * viewpoint). NUNCA lo re-genera: la fuente única es el service (regla de
 * dependencias del PLAN §1). Parser sobre DOMParser (navegador y jsdom).
 */

/** Terna XYZ en coordenadas IFC (Z-up), como las escribe el BCF. */
export type BcfVec3 = [number, number, number];

/** Topic de markup.bcf (subconjunto que emite el service — BCF 3.0). */
export interface BcfTopic {
  guid: string;
  title: string;
  type?: string;
  status?: string;
  priority?: string;
  labels: string[];
  creationDate?: string;
  creationAuthor?: string;
  description?: string;
  /** Fichero del viewpoint (p. ej. "viewpoint.bcfv") si el topic lo tiene. */
  viewpointFile?: string;
  viewpointGuid?: string;
}

/** PerspectiveCamera del viewpoint (D29), en coordenadas IFC (Z-up). */
export interface BcfCamera {
  viewPoint: BcfVec3;
  direction: BcfVec3;
  up: BcfVec3;
  fovDeg: number;
  aspect: number;
}

/** viewpoint.bcfv: selección de componentes + cámara opcional. */
export interface BcfViewpoint {
  guid: string;
  /** IfcGuid de los Components/Selection (los elementos del topic). */
  selection: string[];
  camera?: BcfCamera;
}

function xml(texto: string): Document {
  const doc = new DOMParser().parseFromString(texto, "text/xml");
  const err = doc.querySelector("parsererror");
  if (err) throw new Error(`BCF: XML mal formado — ${err.textContent ?? ""}`);
  return doc;
}

function texto(padre: Element, tag: string): string | undefined {
  for (const el of Array.from(padre.children)) {
    if (el.tagName === tag) return el.textContent ?? undefined;
  }
  return undefined;
}

/** Parsea un markup.bcf (BCF 3.0) → topic. Lanza si no hay <Topic>. */
export function parseMarkup(textoXml: string): BcfTopic {
  const doc = xml(textoXml);
  const topic = doc.querySelector("Markup > Topic");
  if (!topic) throw new Error("BCF: markup.bcf sin <Topic>");
  const labels: string[] = [];
  const labelsEl = Array.from(topic.children).find((e) => e.tagName === "Labels");
  if (labelsEl) {
    for (const l of Array.from(labelsEl.children)) {
      if (l.tagName === "Label" && l.textContent) labels.push(l.textContent);
    }
  }
  const out: BcfTopic = {
    guid: topic.getAttribute("Guid") ?? "",
    title: texto(topic, "Title") ?? "(sin título)",
    labels,
  };
  const type = topic.getAttribute("TopicType");
  if (type) out.type = type;
  const status = topic.getAttribute("TopicStatus");
  if (status) out.status = status;
  const priority = texto(topic, "Priority");
  if (priority) out.priority = priority;
  const creationDate = texto(topic, "CreationDate");
  if (creationDate) out.creationDate = creationDate;
  const creationAuthor = texto(topic, "CreationAuthor");
  if (creationAuthor) out.creationAuthor = creationAuthor;
  const description = texto(topic, "Description");
  if (description) out.description = description;
  const vp = topic.querySelector("Viewpoints > ViewPoint");
  if (vp) {
    const vpGuid = vp.getAttribute("Guid");
    if (vpGuid) out.viewpointGuid = vpGuid;
    const vpFile = texto(vp, "Viewpoint");
    if (vpFile) out.viewpointFile = vpFile;
  }
  return out;
}

function terna(padre: Element, tag: string): BcfVec3 {
  const el = Array.from(padre.children).find((e) => e.tagName === tag);
  if (!el) throw new Error(`BCF: PerspectiveCamera sin <${tag}>`);
  const v = (n: string): number => {
    const t = texto(el, n);
    if (t === undefined) throw new Error(`BCF: <${tag}> sin <${n}>`);
    return Number(t);
  };
  return [v("X"), v("Y"), v("Z")];
}

/** Parsea un viewpoint.bcfv (BCF 3.0) → selección + cámara (si la hay). */
export function parseViewpoint(textoXml: string): BcfViewpoint {
  const doc = xml(textoXml);
  const root = doc.querySelector("VisualizationInfo");
  if (!root) throw new Error("BCF: viewpoint.bcfv sin <VisualizationInfo>");
  const selection: string[] = [];
  for (const c of Array.from(root.querySelectorAll("Components > Selection > Component"))) {
    const g = c.getAttribute("IfcGuid");
    if (g) selection.push(g);
  }
  const out: BcfViewpoint = { guid: root.getAttribute("Guid") ?? "", selection };
  const cam = root.querySelector("PerspectiveCamera");
  if (cam) {
    out.camera = {
      viewPoint: terna(cam, "CameraViewPoint"),
      direction: terna(cam, "CameraDirection"),
      up: terna(cam, "CameraUpVector"),
      fovDeg: Number(texto(cam, "FieldOfView") ?? "60"),
      aspect: Number(texto(cam, "AspectRatio") ?? "1"),
    };
  }
  return out;
}

/** Mapea la cámara BCF (coordenadas IFC, Z-up) al marco del visor (Y-up de
 *  web-ifc/three): (x, y, z)_IFC → (x, z, −y)_visor. Es el mismo cambio de eje
 *  que aplica web-ifc a la geometría, así la cámara cae donde los meshes. */
export function bcfCameraToViewer(c: BcfCamera): {
  position: BcfVec3; direction: BcfVec3; up: BcfVec3; fovDeg: number;
} {
  const m = (v: BcfVec3): BcfVec3 => [v[0], v[2], -v[1]];
  return { position: m(c.viewPoint), direction: m(c.direction), up: m(c.up), fovDeg: c.fovDeg };
}
